Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -0,0 +1,828 @@
+/-
+Copyright (c) 2023 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.Point
+
+/-!
+# Projective coordinates of Weierstrass curves
+-/
+
+abbrev WeierstrassCurve.Projective :=
+  WeierstrassCurve
+
+universe u
+
+namespace WeierstrassCurve.Projective
+
+open MvPolynomial
+
+local macro "eval_simp" : tactic =>
+  `(tactic| simp only [eval_C, eval_X, eval_add, eval_sub, eval_mul, eval_pow])
+
+structure PointRep (R : Type u) where
+  (x y z : R)
```

Comment:
The implementation I had in mind is to define the projective space as a subtype (or just a subset) of a quotient type of `Fin 3 → R`, and define `PointRep R` as a subtype of that (or just a smaller subset of the quotient type). Doesn't `Fin 3` make it easier to connect to MvPolynomial? I think my proposal might be more broadly applicable and convenient, but you have more experience working on this to decide.

Reply:
Sorry, it seems your implementation is pretty close to what I have in mind, just that you're using `(x y z : R)` rather than `Fin 3 → R`. But I think `PointRel` and general APIs (e.g. reduction) should be defined in terms of tuples indexed by an arbitrary type (`Fin 3` here) to be more widely applicable.

Reply:
Yeah, I wrote and rewrote this about seven times and I did have `Fin 3 → R` originally, but here are two possible issues:
- I would have to write a `Setoid` instance for `Fin 3 → R` that is visible globally (e.g. for use in other files downstream), which might pose some problems for other people who don't want this instance (e.g. when I want the analogous thing for weighted projective space). Maybe this is a non-issue but I'm not sure.
- At the beginning I had the issue that `P` is not defeq to `![P.0, P.1, P.2]` (whereas my `P` is defeq to `<P.x, P.y, P.z>`), and this is slightly annoying when trying to define some predicates and structures in term mode (not a problem for proofs). I have since rewritten the entire file several times so maybe this is no longer problematic.

If you think these aren't big issues, I can try making it work for `Fin 3 → R` again.

Reply:
> I would have to write a `Setoid` instance for `Fin 3 → R` that is visible globally (e.g. for use in other files downstream), which might pose some problems for other people who don't want this instance (e.g. when I want the analogous thing for weighted projective space). Maybe this is a non-issue but I'm not sure.

The usual approach is make the `instSetoidPointRep` a `def` and write `attribute [local instance]` when you need to activate it.

> At the beginning I had the issue that `P` is not defeq to `![P.0, P.1, P.2]` (whereas my `P` is defeq to `<P.x, P.y, P.z>`), and this is slightly annoying when trying to define some predicates and structures in term mode (not a problem for proofs). I have since rewritten the entire file several times so maybe this is no longer problematic.

I don't quite get what the problem in term mode is and would appreciate an example. And what are the "structures" that you mention? In any case, wouldn't you define predicates in terms of `P 0`, `P 1` and `P 2` the same way you define them in terms of `P.x`, `P.y` and `P.z`? You then need to supply a proof it's well-defined to define predicates on the quotient type, but you can do it in tactic mode.

Reply:
Fair enough. I don't have an example in mind because this was a while ago. Let me try this again.

Reply:
Just realized: PointRel can be defined on any R-module (not just `ι → R`) and is actually already available as the [MulAction.orbitRel](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/GroupAction/Basic.html#MulAction.orbitRel) of [Units.instMulActionUnitsToMonoidToDivInvMonoidInstGroupUnits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/GroupAction/Units.html#Units.instMulActionUnitsToMonoidToDivInvMonoidInstGroupUnits), so we can just define `PointClass` as [MulAction.orbitRel.Quotient](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/GroupAction/Basic.html#MulAction.orbitRel.Quotient).

For modules of the form `ι → R` we can show the ideal generated by the coordinates is invariant under the action, and define the projective space as the subtype where this ideal is the unit ideal. I can't find this in your code, maybe because you don't need this. (Because it's implied by the nonsingularity condition? Maybe we should redefine nonsingularity in terms of the ideal generated by the partial derivatives for compatibility with the projective space and smoothness.)

Reply:
I replaced `PointRep` with `Fin 3 → R`, but now there will be lots of `matrix_simp`'s and `smul_fin3`'s sprinkled in places.

Yes it's implied by the nonsingularity condition. Why do you need the predicates to be defined in terms of ideal generated by the coordinates or partial derivatives? Is this because your base ring is not a field?

Reply:
> I replaced `PointRep` with `Fin 3 → R`, but now there will be lots of `matrix_simp`'s and `smul_fin3`'s sprinkled in places.

Thanks! The diff of [a839ab](https://github.com/leanprover-community/mathlib4/commit/a839abc4702bd47c36f4cba95a86dcbe96f422b2) looks pretty reasonable to me (+394 -397). When I'm looking at it I realize that `P 0` `P 1` `P 2` etc. is more obscure than the old `P.x` `P.y` `P.z`, and I came up with some potential tricks to use notations similar to the old ones:

We may introduce a `abbrev axis := Fin 3` and use `axis → R` instead of `Fin 3 → R`. Then we define `def axis.x : axis := 0`, `def axis.y : axis := 1` etc. so that for `P : axis → R` we may write `P .x` etc. Or maybe we can just introduce `private def Fin.x : Fin 3 := 0` etc. without introducing `axis`. We could also introduce `axis` as an inductive type of three constructors `x`, `y`, and `z`, and the advantage would be being able to use `cases` rather than `fin_cases`. (Maybe Lean 4's `cases` can do `fin_cases` as well? Anyway I don't think this is that much an advantage.)

> Yes it's implied by the nonsingularity condition. Why do you need the predicates to be defined in terms of ideal generated by the coordinates or partial derivatives? Is this because your base ring is not a field?

The condition feels more natural as it's preserved by ring homomorphisms and invariant under the action by the group of units, and I think if you define the projective points this way then E(R) maps to E(Frac R) and E(R/m) which identifies E(R) as the nonsingular points of E(Frac R) that reduces to nonsingular points of E(R/m) (for R a DVR, say). And the unit ideal condition on projective points easily reduces to that on the affine points, because of the identity $XF_X+YF_Y+ZF_Z=\text{deg }F\cdot F$ for homogeneous F means that for a point on the curve in the affine chart $Z=1$, $F_Z$ is always contained in the ideal generated by $F_X$ and $F_Y$.

Reply:
Would the current definitions be too troublesome for the maps you need? I would think R -> Frac R and R -> R / m are just `baseChange`s over algebras, and we have the corresponding `baseChange` on the nonsingular points.

I'm trying to give analogous definitions for weighted projective points, and I'm not sure how easy it is to rephrase everything in terms of ideals in this case, and I still need to do the same for the group law, which is why I'm hesitating.

Reply:
Seems like a `local notation` suffices if we just need it to be pretty in the file.